### PR TITLE
Add optional  '--topics' CLI argument for 'ros2 bag record'

### PR DESCRIPTION
**Pull request by MichaelOrlov**
#### Motivation:
- This is a follow-up from the https://github.com/ros2/rosbag2/pull/1629#issuecomment-2084683977
- TLDR:  if one wants to specify a list of services and list of topics for recording, the CLI command should be
```
ros2 bag record --services service1 service2 -- topic1 topic2 --output path/to/bag
```
It works currently, if put a trailing `--` after the list of services. However, that workaround with trailing `--` is not obvious and could cause confusion and be a source of mistakes because the topics parameter can't be specified explicitly. i.e., `--topics`.

#### List of changes:
- Add optional "`--topics`" parameter for the `ros2 bag record` CLI interface.
- Deprecate the positional "`topics`" parameter to be consistent with the"`--services`" parameter.
- Add test coverage for non-trivial cases in the recorder CLI parser.
- Use the optional "`--topics`" parameter instead of the deprecated positional parameter in integration tests.

Note: There are no API/ABI breaking changes in this PR

### CODE REVIEW

File: ros2bag/ros2bag/verb/record.py [hunk 1]
```diff
@@ -34,168 +34,177 @@
...
+    # Topic filter arguments
+    topics_args_group = parser.add_mutually_exclusive_group()
+    topics_args_group.add_argument(
+        'topics_positional', type=str, default=[], metavar='[Topic ...]', nargs='*',
+        help='Space-delimited list of topics to record.')
+    topics_args_group.add_argument(
+        '--topics', type=str, default=[], metavar='Topic', nargs='+',
+        help='Space-delimited list of topics to record.')
+    parser.add_argument(
+        '--topic-types', nargs='+', default=[], help='List of topic types to record.')
```
Comment (Barry-Xu-2018) [inline ros2bag/ros2bag/verb/record.py]: In order to maintain a consistent style of description, how about this type of option should use "space-delimited" as the prompt.
`List of ...` -> `Space-delimited list of`

File: ros2bag/ros2bag/verb/record.py [hunk 2]
```diff
@@ -34,168 +34,177 @@
...
+        help='Record only topics and services containing provided regular expression. '
+             'Overrides --all, --all-topics and --all-services, applies on top of '
+             'topics list and service list.')
+    parser.add_argument(
+        '--exclude-regex', default='',
+        help='Exclude topics and services containing provided regular expression. '
+             'Works on top of --all, --all-topics, or --regex.')
+    parser.add_argument(
+        '--exclude-topic-types', type=str, default=[], metavar='ExcludeTypes', nargs='+',
+        help='List of topic types not being recorded. '
```
Comment (Barry-Xu-2018) [inline ros2bag/ros2bag/verb/record.py]: `List of ...` -> `Space-delimited list of ...`

File: ros2bag/ros2bag/verb/record.py [hunk 3]
```diff
@@ -34,168 +34,177 @@
...
+             'Works on top of --all, --all-topics, or --regex.')
+    parser.add_argument(
+        '--exclude-topics', type=str, metavar='Topic', nargs='+',
+        help='List of topics not being recorded. '
```
Comment (Barry-Xu-2018) [inline ros2bag/ros2bag/verb/record.py]: `List of ...` -> `Space-delimited list of`

File: ros2bag/ros2bag/verb/record.py [hunk 4]
```diff
@@ -34,168 +34,177 @@
...
+             'Works on top of --all, --all-topics, or --regex.')
+    parser.add_argument(
+        '--exclude-services', type=str, metavar='ServiceName', nargs='+',
+        help='List of services not being recorded. '
```
Comment (Barry-Xu-2018) [inline ros2bag/ros2bag/verb/record.py]: `List of ...` -> `Space-delimited list of`

File: ros2bag/ros2bag/verb/record.py [hunk 5]
```diff
@@ -34,168 +34,177 @@
...
+             'Works on top of --all, --all-services, or --regex.')
+
+    # Enable to record service
+    parser.add_argument(
+        '--services', type=str, metavar='ServiceName', nargs='+',
+        help='List of services to record.')
```
Comment (Barry-Xu-2018) [inline ros2bag/ros2bag/verb/record.py]: `List of ...` -> `Space-delimited list of`

File: ros2bag/ros2bag/verb/record.py [hunk 6]
```diff
@@ -208,6 +217,11 @@ def _check_necessary_argument(self, args):
 
     def main(self, *, args):  # noqa: D102
 
+        if args.topics_positional:
```
Comment (Barry-Xu-2018) [inline ros2bag/ros2bag/verb/record.py:221]: Now it is possible to set args.topics_positional and args.topics at the same time.  
For this case, error should be reported.  

```python
if args.topics_positional and args.topics:
    return print_error('...')
```

File: ros2bag/ros2bag/verb/record.py (hunk 1 above)
Comment (MichaelOrlov) [reply ros2bag/ros2bag/verb/record.py]: @Barry-Xu-2018 It's a bit unrelated to the matter of this PR, and I was planning to fix it in the follow-up PR. 
But since you found it and asking for fix, I will fix it in the scope of this PR.
Fixed in the new commit https://github.com/ros2/rosbag2/pull/1632/commits/875e7876b45c676b3ebcd14a69abf402e94aa796

File: ros2bag/ros2bag/verb/record.py (hunk 2 above)
Comment (MichaelOrlov) [reply ros2bag/ros2bag/verb/record.py]: Fixed in the new commit https://github.com/ros2/rosbag2/pull/1632/commits/875e7876b45c676b3ebcd14a69abf402e94aa796

File: ros2bag/ros2bag/verb/record.py (hunk 3 above)
Comment (MichaelOrlov) [reply ros2bag/ros2bag/verb/record.py]: Fixed in the new commit https://github.com/ros2/rosbag2/pull/1632/commits/875e7876b45c676b3ebcd14a69abf402e94aa796

File: ros2bag/ros2bag/verb/record.py (hunk 4 above)
Comment (MichaelOrlov) [reply ros2bag/ros2bag/verb/record.py]: Fixed in the new commit https://github.com/ros2/rosbag2/pull/1632/commits/875e7876b45c676b3ebcd14a69abf402e94aa796

File: ros2bag/ros2bag/verb/record.py (hunk 5 above)
Comment (MichaelOrlov) [reply ros2bag/ros2bag/verb/record.py]: Fixed in the new commit https://github.com/ros2/rosbag2/pull/1632/commits/875e7876b45c676b3ebcd14a69abf402e94aa796

File: ros2bag/ros2bag/verb/record.py (hunk 6 above)
Comment (MichaelOrlov) [reply ros2bag/ros2bag/verb/record.py:221]: @Barry-Xu-2018 As regards
> Now it is possible to set args.topics_positional and args.topics at the same time.

That is actually is not possible because `args.topics_positional` and `args.topics` are in mutually exclusive group. See https://github.com/ros2/rosbag2/pull/1632/files#diff-69df89597d099d6eeeb32ecc480c22116dc6c97f32556ef9636eee901f62452cR62-R68 . The Python arguments parser will detect such situation if one will try to combine a positional argument with optional `--topics` and throw an error with the error message about it. I have checked it.

Comment (Barry-Xu-2018) [reply ros2bag/ros2bag/verb/record.py:221]: Sorry. I didn't notice this line `topics_args_group = parser.add_mutually_exclusive_group()`.

**Comment by Barry-Xu-2018** [comment]
LGTM

**Comment by MichaelOrlov** [comment]
@clalancette @emersonknapp I need a formal review/approval for this PR.

File: ros2bag/ros2bag/verb/record.py [hunk 7]
```diff
@@ -34,168 +34,178 @@
...
+             'defaults to a timestamped folder in the current directory.')
+    parser.add_argument(
+        '-s', '--storage', default=default_writer, choices=writer_choices,
+        help="Storage identifier to be used, defaults to '%(default)s'.")
+
+    # Topic filter arguments
+    topics_args_group = parser.add_mutually_exclusive_group()
+    topics_args_group.add_argument(
+        'topics_positional', type=str, default=[], metavar='[Topic ...]', nargs='*',
+        help='Space-delimited list of topics to record.')
```
Comment (fujitatomoya) [inline ros2bag/ros2bag/verb/record.py]: maybe we can add deprecation notice in help message as well?

```suggestion
        help='Space-delimited list of topics to record. (deprecated)')
```

**Review by fujitatomoya — APPROVED** [review]
@MichaelOrlov lgtm with minor comment. thanks!

File: ros2bag/ros2bag/verb/record.py (hunk 7 above)
Comment (MichaelOrlov) [reply ros2bag/ros2bag/verb/record.py]: Added

**Comment by MichaelOrlov** [comment]
Gist: https://gist.githubusercontent.com/MichaelOrlov/dc71cf596e1d88b0a21fc5b572384bd1/raw/b77faf036884f4558a559c99f6f6d8c6c4c8f617/ros2.repos
BUILD args:  --packages-above-and-dependencies ros2bag rosbag2_tests
TEST args:  --packages-above ros2bag rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/13819
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20999/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15301/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21678/)

**Comment by MichaelOrlov** [comment]
@clalancette @emersonknapp Friendly ping here for formal review/approval.

**Review by emersonknapp — APPROVED** [review]
haven't read all the way through but I'm now coming off my parental leave and trying to get back into the ros2 development. this looks fine at a glance to me

**Comment by MichaelOrlov** [comment]
https://github.com/Mergifyio backport jazzy
